Documentation/error-handling.md

```diff
@@ -0,0 +1,150 @@
+
+### Using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`
+
+The gRPC API provides several options that can be used to configure the behavior
+of dialing and connecting to a gRPC server. Some of these options, such as
+`FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`, rely on
+failures at dial time. However, we strongly discourage developers from using
+these options, as they can introduce race conditions and result in unreliable
+and difficult-to-debug code.
+
+###  Why we discourage using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`
+
+When a client attempts to connect to a gRPC server, it can encounter a variety
+of errors, including network connectivity issues, server-side errors, and
+incorrect usage of the gRPC API. The options `FailOnNonTempDialError`,
+`WithBlock`, and `WithReturnConnectionError` are designed to handle some of
+these errors, but they do so by relying on failures at dial time. This means
+that they may not provide reliable or accurate information about the status of
+the connection.
+
+For example, if a client uses `WithBlock` to wait for a connection to be
+established, it may end up waiting indefinitely if the server is not responding.
+Similarly, if a client uses `WithReturnConnectionError` to return a connection
+error if dialing fails, it may miss opportunities to recover from transient
+network issues that are resolved shortly after the initial dial attempt.
+
+## Best practices for error handling in gRPC
+
+Instead of relying on failures at dial time, we strongly encourage developers to
+rely on errors from RPCs. When a client makes an RPC, it can receive an error
+response from the server. These errors can provide valuable information about
+what went wrong, including information about network issues, server-side errors,
+and incorrect usage of the gRPC API.
+
+By handling errors from RPCs correctly, developers can write more reliable and
+robust gRPC applications. Here are some best practices for error handling in
+gRPC:
+
+- Always check for error responses from RPCs and handle them appropriately.  
+- Use the `status` field of the error response to determine the type of error that
+occurred.  
+- When retrying failed RPCs, use a backoff strategy to avoid
+overwhelming the server or exacerbating network issues.  
+- Avoid using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`, as these
+options can introduce race conditions and result in unreliable and
+difficult-to-debug code.
+
+## Example: Handling errors from an RPC
+
+The following code snippet demonstrates how to handle errors from an RPC in
+gRPC:
+
+```go 
+ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+defer cancel()
+
+res, err := client.MyRPC(ctx, &MyRequest{}) if err != nil {
+// Handle the error appropriately, e.g. log it, return an error to the caller, etc.
+log.Printf("Error calling MyRPC: %v", err) return nil, err
```

Comment:
This needs to be two lines. And please indent code appropriately, so it renders nicely.